File: RemixedDungeon/src/main/java/com/nyrds/pixeldungeon/mobs/elementals/WaterElemental.java
package com.nyrds.pixeldungeon.mobs.elementals;

import com.nyrds.pixeldungeon.mobs.common.IDepthAdjustable;
import com.nyrds.pixeldungeon.mobs.common.MultiKindMob;
import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.blobs.Freezing;
import com.watabou.pixeldungeon.actors.buffs.Bleeding;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Burning;
import com.watabou.pixeldungeon.actors.buffs.Frost;
import com.watabou.pixeldungeon.items.potions.PotionOfFrost;
import com.watabou.pixeldungeon.items.scrolls.ScrollOfPsionicBlast;
import com.watabou.pixeldungeon.levels.TerrainFlags;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class WaterElemental extends MultiKindMob implements IDepthAdjustable {

	public WaterElemental() {
		adjustStats(Dungeon.depth);
		
		loot(new PotionOfFrost(), 0.1f);
	}
	
	public void adjustStats(int depth) {
		kind = Math.min(depth/5, 4);
		if ( kind > 5 ) { kind = 5; }

		hp(ht(depth * 5 + 1));
		baseDefenseSkill = depth * 2 + 1;
		expForKill = depth + 1;
		maxLvl = depth + 2;
		dr = expForKill /3;
		baseAttackSkill = baseDefenseSkill / 2 + 1;
		dmgMin = ht()/2;
		dmgMax = ht()/2;


		addImmunity( Frost.class );
		addImmunity( ScrollOfPsionicBlast.class );
		addImmunity( Bleeding.class );
	}

	@Override
	public int attackProc(@NotNull Char enemy, int damage ) {
		if (Random.Int( 2 ) == 0) {
			Freezing.affect( enemy.getPos() );
		}
		
		return damage;
	}

	@Override
	public float speed() {
		if(TerrainFlags.is(level().map[getPos()], TerrainFlags.LIQUID)) {
			return super.speed() * 2f;
		}

		return super.speed() * 0.5f;
	}

	@Override
	public boolean act() {
		if (level().water[getPos()]) {
			heal(expForKill,this);
		}
		
		return super.act();
	}
	
	@Override
	public boolean add(Buff buff ) {
		if (buff instanceof Frost) {
			if (hp() < ht()) {
				heal(expForKill, buff);
			}
			return false;
		} else {
			if(!Dungeon.isLoading()) {
				if (buff instanceof Burning) {
					damage(Random.NormalIntRange(1, ht() / 3), buff);
				}
			}
			return super.add( buff );
		}
	}
}

File: scripts/mobs/WaterElemental.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

-- java WaterElemental: fast in water, sluggish on land; heals while standing
-- in water; hits chill; Frost is absorbed (heal + no attach), Burning hurts
-- it yet still attaches. Stats scale with the spawn depth (json holds the
-- depth-1 defaults), the depth it lives on is the only depth it ever sees.
return mob.init{
    stats = function(self)
        local d = RPD.Dungeon.depth
        local ht = d * 5 + 1
        self:ht(ht)
        self:hp(ht)
        self:STR(14)
        self:setBaseDefenseSkill(d * 2 + 1)
        self:setBaseAttackSkill(math.floor((d * 2 + 1) / 2) + 1)
        self:setExpForKill(d + 1)
        self:setMaxLvl(d + 2)
        self:setDr(math.floor((d + 1) / 3))
        self:setDmgMin(math.floor(ht / 2))
        self:setDmgMax(math.floor(ht / 2))
    end,

    speed = function(self, base)
        if RPD.TerrainFlags:is(RPD.Dungeon.level.map[self:getPos() + 1], RPD.TerrainFlags.LIQUID) then
            return base * 2
        end
        return base * 0.5
    end,

    act = function(self)
        if RPD.Dungeon.level.water[self:getPos() + 1] then
            self:heal(self:getExpForKill(), self)
        end
    end,

    attackProc = function(self, enemy, dmg)
        -- java hit proc: Freezing.affect on the victim cell - Frost (with
        -- duration), fire put out, and the heap under the victim freezes
        -- (meat -> FrozenCarpaccio, potions shatter, mimics wake chilled).
        -- The Frost itself still shatters to the same hit's damage.
        if enemy ~= nil and math.random(2) == 1 then
            RPD.PseudoBlobs.Freezing:affect(enemy:getPos())
        end
        return dmg
    end,

    addBuff = function(self, buff)
        local kind = buff:getEntityKind()
        if kind == "Frost" then
            -- absorb: heal (src = self, else resist() zeroes a heal sourced
            -- by the Frost itself) and don't attach. The json no longer
            -- lists Frost as immunity so the buff actually reaches here.
            if self:hp() < self:ht() then
                self:heal(self:getExpForKill(), self)
            end
            return true
        end
        if kind == "Burning" then
            if not RPD.Dungeon:isLoading() then
                self:damage(RPD.Random:NormalIntRange(1, math.floor(self:ht() / 3)), buff)
            end
            return false
        end
        return false
    end,
}
